extra_tests/snippets/builtin_exceptions.py

```diff
@@ -0,0 +1,151 @@
+    assert isinstance(exc, MyError)
+    assert exc.__cause__ is e
+    assert exc.__context__ is None
+else:
+    assert False, 'exception not raised'
+
+try:
+    raise ValueError('test') from e
+except ValueError as exc:
+    sys.excepthook(type(exc), exc, exc.__traceback__)  # ok, will print two excs
+    assert isinstance(exc, ValueError)
+    assert exc.__cause__ is e
+    assert exc.__context__ is None
+else:
+    assert False, 'exception not raised'
+
+
+# New case:
+# potential recursion on `__context__` field
+
+e = MyError('message')
+
+try:
+    try:
+        raise e
+    except MyError as exc:
+        raise e
+    else:
+        assert False, 'exception not raised'
+except MyError as exc:
+    sys.excepthook(type(exc), exc, exc.__traceback__)
+    assert exc.__cause__ is None
+    assert exc.__context__ is None
+else:
+    assert False, 'exception not raised'
+
+e = MyError('message')
+
+try:
+    try:
+        raise e
+    except MyError as exc:
+        raise exc
+    else:
+        assert False, 'exception not raised'
+except MyError as exc:
+    sys.excepthook(type(exc), exc, exc.__traceback__)
+    assert exc.__cause__ is None
+    assert exc.__context__ is None
+else:
+    assert False, 'exception not raised'
+
+e = MyError('message')
+
+try:
+    try:
+        raise e
+    except MyError as exc:
+        raise e from e
+    else:
+        assert False, 'exception not raised'
+except MyError as exc:
+    sys.excepthook(type(exc), exc, exc.__traceback__)
+    assert exc.__cause__ is e
+    assert exc.__context__ is None
+else:
+    assert False, 'exception not raised'
+
+e = MyError('message')
+
+try:
+    try:
+        raise e
+    except MyError as exc:
+        raise exc from e
+    else:
+        assert False, 'exception not raised'
+except MyError as exc:
+    sys.excepthook(type(exc), exc, exc.__traceback__)
+    assert exc.__cause__ is e
+    assert exc.__context__ is None
+else:
+    assert False, 'exception not raised'
+
+
+# New case:
+# two exception in a recursion loop
+
+class SubError(MyError):
+    pass
+
+e = MyError('message')
+d = SubError('sub')
+
+
+try:
+    raise e from d
+except MyError as exc:
+    # It was a segmentation fault before, will print info to stdout:
+    sys.excepthook(type(exc), exc, exc.__traceback__)
+    assert isinstance(exc, MyError)
+    assert exc.__cause__ is d
+    assert exc.__context__ is None
+else:
+    assert False, 'exception not raised'
+
+e = MyError('message')
+
+try:
+    raise d from e
+except SubError as exc:
+    # It was a segmentation fault before, will print info to stdout:
+    sys.excepthook(type(exc), exc, exc.__traceback__)
+    assert isinstance(exc, SubError)
+    assert exc.__cause__ is e
+    assert exc.__context__ is None
+else:
+    assert False, 'exception not raised'
+
+
+# New case:
+# explicit `__context__` manipulation.
+
+e = MyError('message')
+e.__context__ = e
+
+try:
+    raise e
+except MyError as exc:
```

Comment:
I think its actually CPython hanging on this (not sure why CPython runs these snippets) probably due to some ciruclar reference clean-up issues? Not really sure. Running this snippet of code with a build of rustpython passes. 

Changing this to `except Exception: pass` should make the check pass.

Reply:
Yes, I can confirm that `CPython` hangs, while `RustPython` works fine.
I will research if we need to submit this bug to CPython team.

Is there any way to distinguish `RustPython` from `CPython` in runtime?

Reply:
I see `python_implementation`

Reply:
@DimitrisJim it is already reported https://bugs.python.org/issue25782

Reply:
Good find. That issue is slightly different in scope as far as I can tell (not really involving printing the exceptions). Rustpython also hangs on some of the cycles they create but that should be tackled in a different issue.